File: haskell/glue/src/Glue/Lib/List/Butlast.hs
module Glue.Lib.List.Butlast where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

butlast :: IR Eval
butlast = NativeFunc butlastImpl

butlastImpl :: IR Eval -> Eval (IR Eval)
butlastImpl arg = case arg of
    List [] -> throwError $ wrongArgumentType ["non-empty list"]
    List [_] -> pure $ List []
    List xs -> pure $ List (init xs)
    _ -> throwError $ wrongArgumentType ["list"]

File: dart/glue/lib/src/lib/list/butlast.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Butlast function - returns all elements except the last one
/// Mirrors Haskell Glue.Lib.List.Butlast.butlast exactly
Ir butlast = IrNativeFunc(butlastImpl);

/// Butlast function implementation
/// Mirrors Haskell Glue.Lib.List.Butlast.butlastImpl exactly
Eval<Ir> butlastImpl(Ir arg) {
  return switch (arg) {
    IrList(elements: final elements) =>
      elements.isEmpty
          ? throwError(wrongArgumentType(['non-empty list']))
          : elements.length == 1
          ? Eval.pure(IrList([]))
          : Eval.pure(
              IrList(elements.sublist(0, elements.length - 1).toList()),
            ),
    _ => throwError(wrongArgumentType(['list'])),
  };
}
